### PR TITLE
Added virtual gpio driver

### PR DESCRIPTION
**Pull request by HumJ0218**
WIP, please make suggestions for amendments here

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1823)

### CODE REVIEW

**Comment by pgrawehr** [comment]
Please add a description of the use case for this. We already have a `DummyGpioDriver` that does basically the same (that is nothing, actually).

**Comment by HumJ0218** [comment]
@pgrawehr  It is a virtual GpioController. (Whether the driver and controller need to be independent of each other is yet to be discussed)

An `Input` method is provided to simulate GPIO pin input changes. An `OutputPinValueChanged` event triggered when the GpioController writes a pin value.

The purpose of this controller is to detach from the hardware device and use pure software program to simulate a real GPIO device. Similar to a simulation program, it helps to save development costs.

**Comment by HumJ0218** [comment]
I recorded a demo video: https://www.bilibili.com/video/BV1Aa411h7Mp

**Comment by Ellerbach** [comment]
@HumJ0218 I guess that it would be perfect to add a README and sme usage and a use case to make it clear.

**Comment by pgrawehr** [comment]
Consider looking into https://github.com/mcNets/Gpio.Driver.Simulator

**Comment by pgrawehr** [comment]
@HumJ0218 That video looks cool, actually. Would it be possible to also include some of the client code you use there? I think this would be really useful if one could control the virtual pins from an external tool (or a browser), as you seem to be doing there.

File: src/devices/VirtualGpio/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Virtual GPIO Controller
+
+As the name says, `VirtualGpio` is a virtualized GPIO controller. It based on `GpioController` and defines an input method.
+
+It allows you to control input pins of GPIO using software code without having to rely on specific hardware. For simulation, testing, etc.
+
+# Usage
+
+## Create vGPIO controller
```
Comment (Ellerbach) [inline src/devices/VirtualGpio/README.md:9]: You need a new line after a title:

```suggestion
## Create vGPIO controller

```

File: src/devices/VirtualGpio/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+```csharp
+VirtualGpioController vGpio = VirtualGpioController.Create(16);
+```
+
+## Add event handlers
```
Comment (Ellerbach) [inline src/devices/VirtualGpio/README.md:14]:
```suggestion
## Add event handlers

```

File: src/devices/VirtualGpio/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+PinChangeEventHandler pinValueChanged = (sender, args) =>
+{
+    Console.WriteLine($"PinNumber: {args.PinNumber}, ChangeType: {args.ChangeType}");
+};
+
+vGpio.OutputPinValueChanged += pinValueChanged;
+vGpio.InputPinValueChanged += pinValueChanged;
+```
+
+## Write output pins
```
Comment (Ellerbach) [inline src/devices/VirtualGpio/README.md:25]:
```suggestion
## Write output pins

```

File: src/devices/VirtualGpio/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+```csharp
+vGpio.OpenPin(0, PinMode.Output);
+vGpio.Write(0, PinValue.High); // set to high -> trigger `OutputPinValueChanged` event
+```
+
+## Control input pins
```
Comment (Ellerbach) [inline src/devices/VirtualGpio/README.md:31]:
```suggestion
## Control input pins

```

File: src/devices/VirtualGpio/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+```csharp
+vGpio.Input(0, PinValue.Low); // connect to low -> shorted, throw exception
+```
```
Comment (Ellerbach) [inline src/devices/VirtualGpio/README.md:34]: You need to add text between the code blocks. Also please respect the new lines you need to have.

File: src/devices/VirtualGpio/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+```csharp
+vGpio.OpenPin(1, PinMode.Input);
+vGpio.Input(1, PinValue.High);  // connect to high -> trigger `InputPinValueChanged` event
+vGpio.Input(1, null); // input Hi-Z(disconnect) -> ndefined behavior -> stay high -> no event
+```
```
Comment (Ellerbach) [inline src/devices/VirtualGpio/README.md:39]: Same here

File: src/devices/VirtualGpio/README.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+As the name says, `VirtualGpio` is a virtualized GPIO controller. It based on `GpioController` and defines an input method.
+
+It allows you to control input pins of GPIO using software code without having to rely on specific hardware. For simulation, testing, etc.
+
+# Usage
+
+## Create vGPIO controller
+```csharp
+VirtualGpioController vGpio = VirtualGpioController.Create(16);
```
Comment (Ellerbach) [inline src/devices/VirtualGpio/README.md:11]: maybe add a comment on what the 16 means?

**Comment by pgrawehr** [comment]
[Triage] Before we can continue on this one, a design discussion (#1828) is required, as this needs a bit more background than what we have now. If there's no further update, we'll close this for now, until someone finds time to provide an extended design (eventually with a prototype).

**Pull request closed without merging**